### PR TITLE
Deprecate the use of - in identifiers

### LINKED ISSUE #2828: Deprecate the use of - in identifiers
We allow the use of - in identifiers but this creates problems and confusion:

- IDEs all see this is an operation so double clicking on the identifier does not work
- In python it is not supported and in plugins both names meet
- It will conflict with adding arithmetic to the language

Compiler deprecation framework already exists

Ensure we produce a warning of module names as well (in so far this is allowed currently)

### PR DESCRIPTION
**Pull request by FloLey**
# Description

Add a deprecation warning to the compiler when using an identifier that has a '-' in its name.

closes #2828 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Review by wouterdb — APPROVED** [review]

**Review by sanderr — CHANGES_REQUESTED** [review]
Nice work! I think you covered everything but I still need to double check.

File: src/inmanta/ast/statements/define.py [hunk 1]
```diff
@@ -720,6 +762,12 @@ def __init__(self, name: LocatableString, toname: LocatableString) -> None:
         DefinitionStatement.__init__(self)
         self.name = str(name)
         self.toname = str(toname)
+        if "-" in self.toname:
```
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/ast/statements/define.py:765]: currently the import are done this way:
```
def p_import(p: YaccProduction) -> None:
    """import : IMPORT ns_ref"""
    p[0] = DefineImport(str(p[2]), str(p[2]))
    attach_lnr(p, 1)


def p_import_1(p: YaccProduction) -> None:
    """import : IMPORT ns_ref AS ID"""
    p[0] = DefineImport(str(p[2]), p[4])
    attach_lnr(p, 1)

```
So I think only a check on toname is needed if I remove the second str() 
in` p[0] = DefineImport(str(p[2]), str(p[2]))`

[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/ast/statements/define.py:740]: Ok I see, i'll make the change :)

[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/ast/statements/define.py:740]: I must admit I'm a bit conflicted on this. Overall I'm still in favor of raising the error here. Mainly because the error you're referring to is just a consequence of implementation details of `inmanta.module`. It is meant as a value error for a specific method dealing with module requirements, not as a sane report to an end-user compiling a model. Since it just arises from decisions/assumptions made in the project loading, refactors in `inmanta.module` might at any point result in code paths where this error is no longer raised in the case we're talking about. I see the current error more as an indirectly caused internal error that just happens to have a more or less relevant error message. My opinion is we should not depend on internal errors to protect our interfaces.

But like I said, I'm not sure about this myself, and I'm willing to drop this if @arnaudsjs and @wouterdb agree that this is redundant.

[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/ast/statements/define.py:745]: @wouterdb rerequesting your review for this change.

**Review by sanderr — APPROVED** [review]

File: src/inmanta/ast/statements/define.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/ast/statements/define.py:745]: For me it's fine to drop it if the error message is sufficiently clear and helpful to the end-user.

File: src/inmanta/ast/statements/assign.py [hunk 2]
```diff
@@ -274,6 +276,13 @@ def __init__(self, name: str, value: ExpressionStatement) -> None:
         AssignStatement.__init__(self, None, value)
         self.name = name
         self.value = value
+        if "-" in self.name:
+            warning: CompilerDeprecationWarning = CompilerDeprecationWarning(
```
Comment (wouterdb) [inline src/inmanta/ast/statements/assign.py:280]: Why is this one constructed different then the others?

File: src/inmanta/ast/statements/define.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/ast/statements/define.py:745]: 1. I like defense in depth, so the fact that it is checked elsewhere is not an argument to not check it here as well
2. We have to adhere to what is documented: according to the docs, do we allow a `-` in a module name? If not, exception, else, warning.

File: src/inmanta/ast/variables.py [hunk 3]
```diff
@@ -71,7 +71,7 @@ def execute_direct(self, requires: Dict[object, object]) -> object:
     def as_assign(self, value: ExpressionStatement, list_only: bool = False) -> AssignStatement:
         if list_only:
             raise ParserException(self.location, "+=", "Can not perform += on variable %s" % self.name)
-        return Assign(self.name, value)
+        return Assign(self, value)
```
Comment (sanderr) [inline src/inmanta/ast/variables.py:74]: I don't think this is correct. `self` is of type `Reference`, not `LocatableString`. I was thinking you could make `self.name: LocatableString`, then use `str()` everywhere it's currently used. `AttributeReference` seems to violate `Reference's` type constraints though, which could start causing issues if we lose the `str()` cast in this constructor.

File: src/inmanta/ast/statements/assign.py [hunk 4]
```diff
@@ -272,17 +272,12 @@ class Assign(AssignStatement):
     provides:      variable
     """
 
-    def __init__(self, name: str, value: ExpressionStatement) -> None:
+    def __init__(self, variable: LocatableString, value: ExpressionStatement) -> None:
         AssignStatement.__init__(self, None, value)
-        self.name = name
+        self.name = variable.name
```
Comment (sanderr) [inline src/inmanta/ast/statements/assign.py:277]: If you change this you could just as well have `self.name` be a `LocatableString` as well. Just make sure to check where it's currently used and use `str()` where appropriate.

File: src/inmanta/parser/plyInmantaParser.py [hunk 5]
```diff
@@ -19,8 +19,10 @@
 import logging
 import re
 from typing import List, Optional, Union
+from xmlrpc.client import boolean
```
Comment (wouterdb) [inline src/inmanta/parser/plyInmantaParser.py:22]: this is not correct

File: tests/test_parser.py [hunk 6]
```diff
@@ -1818,12 +1848,88 @@ def test_rstring():
...
+    )
+    assert len(statements) == 1
+    assign_stmt = statements[0]
+    assert isinstance(assign_stmt, Assign)
+    assert str(assign_stmt.name) == "a"
+    assert assign_stmt.name.location == Range("test", 2, 1, 2, 2)
+    assert isinstance(assign_stmt.value, StringFormat)
+    attribute_ref = assign_stmt.value.children[0]
+    assert attribute_ref.attribute == "bye"
+    assert attribute_ref.location == Range("test", 2, 8, 2, 26)
```
Comment (wouterdb) [inline tests/test_parser.py:1878]: Why is the location of bye not 3 characters long?

**Review by wouterdb — CHANGES_REQUESTED** [review]
Very cool, a bit more than expected but definitely an improvement.

File: tests/test_parser.py (hunk 6 above)
Comment (FloLey) [reply tests/test_parser.py:1878]: Currently I create one Range for the string `{{hello.world.bye}} `and it is used both for the references as for the attributes. I could try to refine it more but I found that the code already became a lot harder to read just to create this LocatableStrings.
https://github.com/inmanta/inmanta-core/blob/583ce25bd5b332c3146b48efd6640cd5c2468e36/src/inmanta/parser/plyInmantaParser.py#L751-L761
I can try going even further if you think the current changes does not impact readability to much and if exact locations are needed.

Comment (FloLey) [reply tests/test_parser.py:1878]: @sanderr

File: src/inmanta/parser/plyInmantaParser.py [hunk 7]
```diff
@@ -731,40 +733,52 @@ def p_constant_rstring(p: YaccProduction) -> None:
...
+    locatable_matches: List(LocatableString) = []
+    for obj in matches:
+        init, match, _ = (str(string)[start:end].splitlines() for start, end in pairwise((0, obj[1], obj[2], len(obj))))
+        start_line: int = string.lnr + len(init) - 1
+        end_line: int = start_line + len(match) - 1
+        len_last_init_line: int = len(init[-1]) if init else 0
+        start_char: int = len(init[-1]) + offset + string.start if len(init) == 1 else len_last_init_line + 1
+        end_char: int = len(match[-1]) + start_char - 1 if len(match) == 1 else len(match[-1])
+        range: Range = Range(string.location.file, start_line, start_char, end_line, end_char)
+        locatable_string = LocatableString("".join(match), range, string.lexpos, string.namespace)
```
Comment (sanderr) [inline src/inmanta/parser/plyInmantaParser.py:760]: `"".join(match)` is a simplification of the actual format string. At first I thought it wouldn't matter at all but when looking into `StringFormat.execute` it seems that it expects the variables as passed in its constructor to appear exactly like that in the string. Meaning that if any newlines occur between the `{{` and `}}`, this won't work. `"\n".join(match)` would not be sufficient either because of potentially different line endings (`\r\n` for example).

So, `create_string_format` requires the exact string (`obj[0]` I think). There are a couple of different approaches of course. My suggestion would be to combine this with my comment regarding using the `Match` object (can't link there for some reason). You could then easily have this whole loop construct a `LocatableString` just for the variable part, and then pass  a tuple of that and the exact substring (`match.group(0)`) to `create_string_format`.

File: src/inmanta/parser/plyInmantaParser.py [hunk 8]
```diff
@@ -731,40 +733,52 @@ def p_constant_rstring(p: YaccProduction) -> None:
...
     _vars = []
-
-    for var_str in variables:
-        var_parts = var_str[1].split(".")
-        ref = Reference(var_parts[0])
-        ref.namespace = namespace
-
+    for var in variables:
+        var_parts: List[str] = str(var)[2:-2].strip().split(".")
+        ref_locatable_string = LocatableString(var_parts[0], var.location, var.lexpos, var.namespace)
```
Comment (sanderr) [inline src/inmanta/parser/plyInmantaParser.py:772]: I see what you meant now. I would prefer this (and the `AttributeReference` one) to be accurate but I don't think it's worth too much effort. That said, I think getting the correct range should be quite a lot simpler here than for the loop above. Something like the snippet below could get you the required offset for each attribute:
```python
attribute_offsets: Iterator[int] = accumulate(var_parts[1:], lambda acc, part: acc + len(part) + 1, initial=len(var_parts[0] + 1))
for attr, offset in zip(var_parts[1:], attribute_offsets):
    ...
```

If you discover it turns out to be more complex, I'm fine with keeping this as is for now.

File: src/inmanta/parser/plyInmantaParser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py]: yes, I'm not used yet to the name "bool" for boolean and I trusted vscode with the auto import when I wrote "boolean".

File: tests/test_parser.py (hunk 6 above)
Comment (sanderr) [reply tests/test_parser.py:1878]: My thoughts [here](https://github.com/inmanta/inmanta-core/pull/3778#discussion_r805711310)

File: src/inmanta/parser/plyInmantaParser.py (hunk 8 above)
Comment (wouterdb) [reply src/inmanta/parser/plyInmantaParser.py:772]: I agree to that

File: src/inmanta/parser/plyInmantaParser.py [hunk 9]
```diff
@@ -731,40 +733,52 @@ def p_constant_rstring(p: YaccProduction) -> None:
...
 
 format_regex = r"""({{\s*([\.A-Za-z0-9_-]+)\s*}})"""
 format_regex_compiled = re.compile(format_regex, re.MULTILINE | re.DOTALL)
 
 
-def get_string_ast_node(string: LocatableString, location: Location) -> Union[Literal, StringFormat]:
-    match_obj = format_regex_compiled.findall(str(string))
-    if len(match_obj) == 0:
+def get_string_ast_node(string: LocatableString, mls: boolean) -> Union[Literal, StringFormat]:
+    matches = [[str(string)[m.start() : m.end()], m.start(), m.end()] for m in format_regex_compiled.finditer(str(string))]
```
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py]: @sanderr  I think the match could be on multiple lines like this:
```
"""{

x.n


}"""
```

[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/parser/plyInmantaParser.py]: I mean if you use `match.group(1)`, this will always be single-line because the group does not allow any whitespace characters.

File: src/inmanta/parser/plyInmantaParser.py (hunk 7 above)
Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py:760]: So by exact string, It would be the string before it gets "splitlined"? can't I just use that as name for the `LocatableString`? and use that in `create_string_format`.
like this:
```
 for obj in matches:
        init_string, match_string = (str(string_ast)[start:end] for start, end in pairwise((0, obj.start(), obj.end())))
        init: List[str] = (init_string + "\n").splitlines()     # lines before the match
        match: List[str] = match_string.splitlines()       # lines matching the string format
        ....
        locatable_string = LocatableString(match_string, range, string_ast.lexpos, string_ast.namespace)
        locatable_matches.append(locatable_string)
 return create_string_format(string_ast, locatable_matches)
        
```

Comment (sanderr) [reply src/inmanta/parser/plyInmantaParser.py:760]: That is one of the options. I think the main advantage of the approach I suggested would be that we immediately have the `Range` for the actual attribute reference instead of the `Range` including `{{ }}` and any whitespace in between. I think that could lead to a cleaner code overall because it would increase accuracy while keeping the code very simple, only at the cost of extending the method to accept a tuple. But if you disagree I'm not against the approach you suggest. Do keep in mind thought that it will require some more advanced logic in `create_string_format` if we want to determine the correct `Range` for the `Reference` and `AttributeReference` values.

File: src/inmanta/parser/plyInmantaParser.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py]: I think I need the other lines later on to compute the location of x and n. To compute the line_offset based on the number of white lines and the start line. I don't see how I could do it otherwise?

File: src/inmanta/ast/variables.py [hunk 10]
```diff
@@ -242,7 +242,7 @@ def __init__(self, instance: Reference, attribute: LocatableString) -> None:
 
         # a reference to the instance
         self.instance = instance
-        self.location = range
+        self.location = attribute.location
```
Comment (wouterdb) [inline src/inmanta/ast/variables.py:245]: why is this changed?

File: src/inmanta/parser/plyInmantaParser.py [hunk 11]
```diff
@@ -736,49 +736,74 @@ def p_constant_mls(p: YaccProduction) -> None:
...
+        return Literal(str(string_ast))
+
+    mls_offset: int = 3 if mls else 1  # len(""")  or len(') or len(")
+    locatable_matches: List[Tuple[str, LocatableString]] = []
+    for match in matches:
+        init_string: str = str(string_ast)[0 : match.start() + 2]
+        match_string: str = match[2]
+        init_lines: List[str] = (init_string + "\n").splitlines()  # lines before the match
+        match_lines: List[str] = match_string.splitlines()  # lines matching the string format
+        line_offset, char_offset = get_offset(match_lines, match[3])
```
Comment (wouterdb) [inline src/inmanta/parser/plyInmantaParser.py:755]: I don't understand this:
1. why match again, if the matching is already done
2. how is this safe against repeated patterns?

This could definitely use some comments.

File: src/inmanta/parser/plyInmantaParser.py [hunk 12]
```diff
@@ -736,49 +736,74 @@ def p_constant_mls(p: YaccProduction) -> None:
...
+        var_parts: List[str] = var_name.split(".")
+        start_char = var.location.start_char
+        end_char = start_char + len(var_parts[0])
+        range: Range = Range(var.location.file, var.location.lnr, start_char, var.location.lnr, end_char)
+        ref_locatable_string = LocatableString(var_parts[0], range, var.lexpos, var.namespace)
         ref = Reference(ref_locatable_string)
         if len(var_parts) > 1:
-            for attr in var_parts[1:]:
-                attr_locatable_string: LocatableString = LocatableString(attr, var.location, var.lexpos, var.namespace)
+            attribute_offsets: Iterator[int] = accumulate(
```
Comment (wouterdb) [inline src/inmanta/parser/plyInmantaParser.py:794]: I've never seen `itertools.accumulate`, nice

File: src/inmanta/parser/plyInmantaParser.py [hunk 13]
```diff
@@ -736,49 +736,74 @@ def p_constant_mls(p: YaccProduction) -> None:
...
+def get_offset(match_lines: List[str], name: str) -> Tuple[int, int]:
+    for line_index in range(len(match_lines)):
+        match: Optional[re.Match[str]] = re.search(name, match_lines[line_index])
+        if match:
+            return line_index, match.start() + 1
+    else:
+        raise Exception("Match not found")
+
+
+def create_string_format(format_string: LocatableString, variables: List[Tuple[str, LocatableString]]) -> StringFormat:
```
Comment (wouterdb) [inline src/inmanta/parser/plyInmantaParser.py:780]: If I read this correctly, it assumes that the 'variables' never contain a newline. 

Do we have a testcase to verify this is effectively not allowed?

File: tests/test_parser.py [hunk 14]
```diff
@@ -326,7 +326,7 @@ def test_new_relation_with_annotations_unidir():
     assert rel.left[2] is None
     assert rel.right[2] == (1, 1)
     assert len(rel.annotations) == 2
-    range1: Range = Range("test", 2, 18, 2, 21)
+    range1: Range = Range("test", 2, 14, 2, 17)
```
Comment (wouterdb) [inline tests/test_parser.py:329]: I assume this was a bug in the equals function?

**Review by wouterdb — COMMENTED** [review]
Impressive work!

File: tests/test_parser.py (hunk 14 above)
Comment (FloLey) [reply tests/test_parser.py:329]: yes, Location had a  __eq__  function that was used by Range as it was not overwritten before.

File: src/inmanta/parser/plyInmantaParser.py (hunk 11 above)
Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py:755]: 1.
lets say we have following string:

```
a = """hello{

      this.is.the.string

}"""
```

The first matching gives use:

```
match[1] = "{{

      this.is.the.string

}}"

```
```
match[2] ="                 #match 1 without the {{}}

      this.is.the.string

"
```


`match[3] = "this.is.the.string" #the actual variables`

With the second matching I look for the position of match[3] inside of match[2]. (knowing the variables should be on the same line)
this is done by:

- Splitting match[2] on linebreaks: ['','',"  `  ``  ``  `this.is.the.string",'']
- Iterating on each element of the array and with a second matching look for match[3] in the element.
- The index where it is found is the number of lines to add to startline of match[2]. and match.start() is the offset caused by the spaces that could be present before the variable.

2.As we only handle one {{}} at the time and it(match[2]) only contains spaces, linebreaks and the match, the match should never be found more then once.

File: src/inmanta/parser/plyInmantaParser.py (hunk 13 above)
Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py:780]: It is indeed assumed. I just tested it I don't think so. even worst it is accepted by the compiler 
```
a ="""{{this.is
.the.string}}"""
std::print(a)
```
will print the string
 ```
"{{this.is
.the.string}}"
```
I'll add a new ticket when this is merged

File: src/inmanta/ast/variables.py (hunk 10 above)
Comment (FloLey) [reply src/inmanta/ast/variables.py:245]: I think this is a mistake, I changed it.

File: src/inmanta/parser/plyInmantaParser.py (hunk 13 above)
Comment (sanderr) [reply src/inmanta/parser/plyInmantaParser.py:780]: I think your example is flawed. The regex does not allow newlines so it is not recognized as a string format, hence it prints the literal string. I do think it's a good idea to add this invariant to the docstring though.

File: tests/test_parser.py [hunk 15]
```diff
@@ -1818,12 +1847,109 @@ def test_rstring():
...
+    assert str(instance2.name) == "hello"
+    assert instance2.name.location == Range("test", 3, 10, 3, 15)
+
+
+def test_string_attribute_reference_4():
+    statements = parse_code(
+        """
+a=\"""
+format string starts as first char on new line
+{{x.n}}
```
[the remark this answers is not in the document]
Comment (FloLey) [reply tests/test_parser.py:1940]: i'll add both :)

File: src/inmanta/parser/plyInmantaParser.py (hunk 11 above)
Comment (sanderr) [reply src/inmanta/parser/plyInmantaParser.py:755]: The reason we didn't is that it isn't necessarily correct: depending on what line endings the file uses you could get incorrect location reporting (`\n`, `\r`, `\r\n`). Since `splitlines` has all the know-how to handle this I thought it would be better to go with that. But I am convinced it can be made almost as readable with `splitlines` as with newline counts.

Comment (wouterdb) [reply src/inmanta/parser/plyInmantaParser.py:755]: @sanderr I think all the code in the parser is built on the assumption newline is '\n' 

```
def t_newline(t: lex.LexToken) -> None:  # noqa: N802
    r"\n+"
```

Comment (wouterdb) [reply src/inmanta/parser/plyInmantaParser.py:756]: @FloLey 
https://docs.python.org/3/library/re.html#re.Match.start

so it is not `match[2].start()` but `match.start(2)`

Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py:756]: Ho this is nice! thanks I'll try it out

Comment (FloLey) [reply src/inmanta/parser/plyInmantaParser.py]: @sanderr @wouterdb I applied the changed removing the splitlines and using the idea of Wouter. It is a lot more readable.
(The tests are passing on the last build, just an issue with flake8 with should be fixed with my last commit)

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/parser/plyInmantaParser.py [hunk 16]
```diff
@@ -745,41 +745,39 @@ def get_string_ast_node(string_ast: LocatableString, mls: bool) -> Union[Literal
     if len(matches) == 0:
         return Literal(str(string_ast))
 
+    start_lnr = string_ast.location.lnr
+    start_char_pos = string_ast.location.start_char
+    whole_string = str(string_ast)
     mls_offset: int = 3 if mls else 1  # len(""")  or len(') or len(")
+
+    def char_count_to_lnr_char(position):
```
Comment (sanderr) [inline src/inmanta/parser/plyInmantaParser.py:753]:
```suggestion
    def char_count_to_lnr_char(position: int) -> Tuple[int, int]:
```

**Review by sanderr — APPROVED** [review]
